### ISSUE TITLE
Bug: Transactions on Oracle are broken

### ISSUE THREAD

**Issue by paya-cz** [knex/knex#4884]
# Environment

Knex version: 0.95.14
Database + version: Oracle 12cR1

@atiertant

# Bug

I have based my code on an example from the documentation:

```ts

try {
    await knex.transaction(
        async trx => {
            const books = [
                { title: 'Canterbury Tales' },
                { title: 'Moby Dick' },
                { title: 'Hamlet' }
            ];

            const ids = await trx('catalogues')
                .insert({
                    name: 'Old Books'
                }, 'id');

            books.forEach((book) => book.catalogue_id = ids[0]);
            const inserts = await trx('books').insert(books);

            console.log(inserts.length + ' new books saved.');
        },
        {
            connection: connection, // connection object from 'oracledb' library
            // isolationLevel: 'serializable', // knexjs does not support this for Oracle
        },
    );
} catch (error) {
    // If we get here, that means that neither the 'Old Books' catalogues insert,
    // nor any of the books inserts will have taken place.
    console.error(error);
}
```

Do notice the extra transaction config object, specifically where I pass my own connection instance.
The line with `const ids = await trx('catalogues')` throws an error:

```
\node_modules\knex\lib\dialects\oracledb\transaction.js:50
     return timeout(conn.rollbackAsync(), 5000)
                         ^
TypeError: conn.rollbackAsync is not a function
     at Oracle_Transaction.rollback
(\node_modules\knex\lib\dialects\oracledb\transaction.js:50:25)
     at Function.transactor.rollback
(\node_modules\knex\lib\execution\transaction.js:315:42)
     at
\node_modules\knex\lib\execution\transaction.js:231:35
```

I assume knex is monkeypatching some functions onto the connection object, and these are missing when the connection obj comes straight out of `oracledb`. I wonder if monkeypatching functions onto foreign objects is the best approach here.

**Comment by cjbland** [comment]
I think this issue is related: https://github.com/knex/knex/issues/4013

**Comment by mercmobily** [comment]
Hi! Thanks for the thorough report and for isolating the external connection case. This is a real bug in the Oracle dialect’s transaction path, not a user mistake.

This took me a while to actually get. I am only triaging for now, but still trying to understand each issue. This one was a little difficult.

Knex’s Oracle dialect decorates `oracledb` connections with small promise helpers (`commitAsync`/`rollbackAsync`/`executeAsync`) and LOB/result‑set handling so transactions integrate cleanly with Knex’s promise/timeout flow. When Knex acquires a connection from its pool, that decoration happens automatically. Likewise, when you attach a user‑supplied connection to an individual query via `.connection(conn)`, Knex calls `client.processPassedConnection(conn)` to decorate it.

The path that fails is the Oracle‑specific transaction code that accepts an external connection via the second argument (for example, `knex.transaction(fn, { connection })`). In that path, the decoration hook was not called. The transaction then tries to invoke `conn.commitAsync`/`conn.rollbackAsync`, which are missing on a raw `oracledb` connection, causing `TypeError: conn.rollbackAsync is not a function`.

Code evidence:

- lib/dialects/oracledb/transaction.js: `commit` uses `conn.commitAsync()` and `rollback` uses `conn.rollbackAsync()`; `acquireConnection` uses `config.connection` directly when supplied.
- lib/dialects/oracledb/utils.js: `monkeyPatchConnection(connection, client)` defines `commitAsync`, `rollbackAsync`, and `executeAsync`.
- lib/dialects/oracledb/index.js: pool‑acquired connections are patched in `acquireRawConnection`; `processPassedConnection(connection)` decorates user‑supplied connections, but that hook is not invoked by the Oracle transaction path that accepts `{ connection }`.

Workarounds right now:

- Prefer letting Knex manage the connection inside a transaction: `await knex.transaction(async (trx) => { /* … */ })`.
- If you must pass an external `oracledb` connection (for example proxy authentication or special session state), decorate it before starting the transaction:

  `const conn = await oracledb.getConnection(...);
   knex.client.processPassedConnection(conn);
   await knex.transaction(async (trx) => { /* … */ }, { connection: conn });`

Fix direction in Knex:

- Call `this.client.processPassedConnection(connection)` (or directly `monkeyPatchConnection(connection, this.client)`) in `Oracle_Transaction.acquireConnection` when `config.connection` is provided, before using it. This mirrors how pool‑acquired connections are prepared and makes external connections work out of the box.
- Optionally introduce a generic hook so any dialect’s `Transaction.acquireConnection` path always runs `client.processPassedConnection` for external connections.

This change will address the exact error you reported. It should be covered by a unit/integration test that starts a transaction with an external `oracledb` connection to prevent regressions. BTW this is not a duplicate of #4013 (different failure mode and root cause); cross‑linking it there is useful for context only.

Related issues:

- #4013 — Prior Oracle transaction issues (CLOSED/older context), but different surface area
